File: export_codebase.js
#!/usr/bin/env node

/**
 * MeetingMind Codebase Export Script (Node.js Version)
 * Exports all code files, structure, and content to a comprehensive text file
 */

const fs = require('fs');
const path = require('path');
const { promisify } = require('util');

const readdir = promisify(fs.readdir);
const stat = promisify(fs.stat);
const readFile = promisify(fs.readFile);
const writeFile = promisify(fs.writeFile);

// Configuration
const ROOT_DIR = __dirname;
const OUTPUT_FILE = path.join(ROOT_DIR, 'complete_codebase_export.txt');

// File patterns to include
const INCLUDE_EXTENSIONS = new Set([
    '.ts', '.tsx', '.js', '.jsx', '.json', '.yaml', '.yml',
    '.py', '.md', '.txt', '.html', '.css', '.scss', '.sass', '.less',
    '.sql', '.sh', '.bat', '.ps1', '.dockerfile',
    '.gitignore', '.gitattributes', '.editorconfig',
    '.lock', '.toml', '.ini', '.cfg', '.properties',
    '.xml', '.svg', '.php', '.rb', '.go', '.rs', '.cpp', '.c', '.h',
    '.java', '.kt', '.swift', '.dart', '.r', '.config', '.conf'
]);

const INCLUDE_FILENAMES = new Set([
    'package.json', 'requirements.txt', 'Pipfile', 'poetry.lock',
    'Cargo.toml', 'go.mod', 'pom.xml', 'build.gradle',
    'dockerfile', 'docker-compose.yml', 'docker-compose.yaml'
]);

// Files to exclude (specific filenames)
const EXCLUDE_FILES = new Set([
    'export_codebase.py', 'export_codebase.js', 'export_codebase.sh',
    'export_codebase.bat', 'EXPORT_README.md', 'complete_codebase_export.txt',
    'complete_codebase.txt'
]);

// Directories to exclude
const EXCLUDE_DIRS = new Set([
    'node_modules', '__pycache__', '.git', '.vscode', '.idea',
    'dist', 'build', 'coverage', '.nyc_output', 'target',
    '.pytest_cache', '.mypy_cache', '.tox', 'venv', 'env',
    '.env', 'virtualenv', '.virtualenv', 'site-packages',
    'vendor', 'bower_components', '.next', '.nuxt', '.cache',
    'tmp', 'temp', '.tmp', '.temp', 'logs', 'log',
    '.DS_Store', 'Thumbs.db', '.sass-cache', '.parcel-cache'
]);

// File extensions to exclude
const EXCLUDE_EXTENSIONS = new Set([
    '.exe', '.dll', '.so', '.dylib', '.bin', '.obj', '.o',
    '.png', '.jpg', '.jpeg', '.gif', '.bmp', '.ico', '.tiff',
    '.mp3', '.mp4', '.wav', '.avi', '.mov', '.wmv', '.flv',
    '.pdf', '.doc', '.docx', '.xls', '.xlsx', '.ppt', '.pptx',
    '.zip', '.tar', '.gz', '.rar', '.7z', '.bz2', '.xz',
    '.woff', '.woff2', '.ttf', '.eot', '.otf',
    '.db', '.sqlite', '.sqlite3', '.mdb', '.accdb'
]);

const MAX_FILE_SIZE = 2 * 1024 * 1024; // 2MB

/**
 * Check if a file should be included
 */
function shouldIncludeFile(filePath, stats) {
    const basename = path.basename(filePath);
    const ext = path.extname(filePath).toLowerCase();
    
    // Check if filename should be excluded
    if (EXCLUDE_FILES.has(basename)) {
        return false;
    }
    
    // Check file size
    if (stats.size > MAX_FILE_SIZE) {
        return false;
    }
    
    // Check if extension should be excluded
    if (EXCLUDE_EXTENSIONS.has(ext)) {
        return false;
    }
    
    // Check if extension should be included
    if (INCLUDE_EXTENSIONS.has(ext)) {
        return true;
    }
    
    // Check if filename should be included
    if (INCLUDE_FILENAMES.has(basename.toLowerCase())) {
        return true;
    }
    
    // Check specific patterns
    if (basename.startsWith('tsconfig') && basename.endsWith('.json')) {
        return true;
    }
    
    if (basename.includes('config') && (ext === '.js' || ext === '.ts' || ext === '.json')) {
        return true;
    }
    
    if (basename.startsWith('.eslintrc') || basename.startsWith('.prettierrc')) {
        return true;
    }
    
    return false;
}

/**
 * Check if a directory should be excluded
 */
function shouldExcludeDir(dirPath) {
    const basename = path.basename(dirPath);
    return EXCLUDE_DIRS.has(basename.toLowerCase());
}

/**
 * Format file size in human readable format
 */
function formatSize(bytes) {
    const units = ['B', 'KB', 'MB', 'GB', 'TB'];
    let size = bytes;
    let unit = 0;
    
    while (size >= 1024 && unit < units.length - 1) {
        size /= 1024;
        unit++;
    }
    
    return `${size.toFixed(1)} ${units[unit]}`;
}

/**
 * Get directory tree structure
 */
async function getDirectoryTree(dirPath, prefix = '', isLast = true) {
    const items = [];
    
    try {
        const entries = await readdir(dirPath);
        const sortedEntries = entries.sort((a, b) => {
            const aPath = path.join(dirPath, a);
            const bPath = path.join(dirPath, b);
            const aIsDir = fs.existsSync(aPath) && fs.statSync(aPath).isDirectory();
            const bIsDir = fs.existsSync(bPath) && fs.statSync(bPath).isDirectory();
            
            if (aIsDir && !bIsDir) return -1;
            if (!aIsDir && bIsDir) return 1;
            return a.localeCompare(b);
        });
        
        for (let i = 0; i < sortedEntries.length; i++) {
            const entry = sortedEntries[i];
            const entryPath = path.join(dirPath, entry);
            
            if (shouldExcludeDir(entryPath)) {
                continue;
            }
            
            const entryIsLast = i === sortedEntries.length - 1;
            const symbol = entryIsLast ? '└── ' : '├── ';
            items.push(`${prefix}${symbol}${entry}`);
            
            try {
                const stats = await stat(entryPath);
                if (stats.isDirectory()) {
                    const extension = entryIsLast ? '    ' : '│   ';
                    const subtree = await getDirectoryTree(entryPath, prefix + extension, entryIsLast);
                    if (subtree.length > 0) {
                        items.push(...subtree);
                    }
                }
            } catch (error) {
                // Skip inaccessible directories
            }
        }
    } catch (error) {
        // Skip inaccessible directories
    }
    
    return items;
}

/**
 * Calculate codebase statistics
 */
async function calculateStats(rootPath) {
    const stats = {
        totalFiles: 0,
        totalLines: 0,
        totalSize: 0,
        fileTypes: {},
        languages: {},
        directories: 0
    };
    
    async function processDirectory(dirPath) {
        try {
            stats.directories++;
            const entries = await readdir(dirPath);
            
            for (const entry of entries) {
                const entryPath = path.join(dirPath, entry);
                
                try {
                    const entryStats = await stat(entryPath);
                    
                    if (entryStats.isDirectory()) {
                        if (!shouldExcludeDir(entryPath)) {
                            await processDirectory(entryPath);
                        }
                    } else if (entryStats.isFile() && shouldIncludeFile(entryPath, entryStats)) {
                        stats.totalFiles++;
                        stats.totalSize += entryStats.size;
                        
                        const ext = path.extname(entryPath).toLowerCase();
                        if (ext) {
                            stats.fileTypes[ext] = (stats.fileTypes[ext] || 0) + 1;
                        } else {
                            stats.fileTypes['no_extension'] = (stats.fileTypes['no_extension'] || 0) + 1;
                        }
                        
                        // Count lines and detect language
                        try {
                            const content = await readFile(entryPath, 'utf8');
                            const lines = content.split('\n').length;
                            stats.totalLines += lines;
                            
                            // Language detection
                            switch (ext) {
                                case '.ts':
                                case '.tsx':
                                    stats.languages['TypeScript'] = (stats.languages['TypeScript'] || 0) + lines;
                                    break;
                                case '.js':
                                case '.jsx':
                                    stats.languages['JavaScript'] = (stats.languages['JavaScript'] || 0) + lines;
                                    break;
                                case '.py':
                                    stats.languages['Python'] = (stats.languages['Python'] || 0) + lines;
                                    break;
                                case '.css':
                                case '.scss':
                                case '.sass':
                                case '.less':
                                    stats.languages['CSS'] = (stats.languages['CSS'] || 0) + lines;
                                    break;
                                case '.html':
                                case '.htm':
                                    stats.languages['HTML'] = (stats.languages['HTML'] || 0) + lines;
                                    break;
                                case '.md':
                                    stats.languages['Markdown'] = (stats.languages['Markdown'] || 0) + lines;
                                    break;
                                case '.json':
                                    stats.languages['JSON'] = (stats.languages['JSON'] || 0) + lines;
                                    break;
                                case '.yaml':
                                case '.yml':
                                    stats.languages['YAML'] = (stats.languages['YAML'] || 0) + lines;
                                    break;
                                default:
                                    stats.languages['Other'] = (stats.languages['Other'] || 0) + lines;
                            }
                        } catch (error) {
                            // Skip files that can't be read as text
                        }
                    }
                } catch (error) {
                    // Skip inaccessible files
                }
            }
        } catch (error) {
            // Skip inaccessible directories
        }
    }
    
    await processDirectory(rootPath);
    return stats;
}

/**
 * Export file content with line numbers
 */
function formatFileContent(content, filePath) {
    const ext = path.extname(filePath).toLowerCase();
    const codeExtensions = ['.ts', '.tsx', '.js', '.jsx', '.py', '.css', '.html', '.json', '.yaml', '.yml'];
    
    if (codeExtensions.includes(ext)) {
        const lines = content.split('\n');
        return lines.map((line, index) => `${(index + 1).toString().padStart(4, ' ')}│ ${line}`).join('\n');
    }
    
    return content;
}

/**
 * Process all files and export content
 */
async function exportFiles(rootPath, outputStream) {
    let exportedCount = 0;
    let skippedCount = 0;
    
    async function processDirectory(dirPath) {
        try {
            const entries = await readdir(dirPath);
            const sortedEntries = entries.sort();
            
            for (const entry of sortedEntries) {
                const entryPath = path.join(dirPath, entry);
                
                try {
                    const entryStats = await stat(entryPath);
                    
                    if (entryStats.isDirectory()) {
                        if (!shouldExcludeDir(entryPath)) {
                            await processDirectory(entryPath);
                        }
                    } else if (entryStats.isFile()) {
                        if (shouldIncludeFile(entryPath, entryStats)) {
                            const relativePath = path.relative(rootPath, entryPath);
                            console.log(`📝 Exporting: ${relativePath}`);
                            
                            // File header
                            outputStream.write('\n' + '='.repeat(80) + '\n');
                            outputStream.write(`FILE: ${relativePath}\n`);
                            outputStream.write(`SIZE: ${formatSize(entryStats.size)}\n`);
                            outputStream.write(`MODIFIED: ${entryStats.mtime.toISOString()}\n`);
                            outputStream.write('='.repeat(80) + '\n\n');
                            
                            // File content
                            try {
                                const content = await readFile(entryPath, 'utf8');
                                const formattedContent = formatFileContent(content, entryPath);
                                outputStream.write(formattedContent);
                                outputStream.write('\n\n');
                                exportedCount++;
                            } catch (error) {
                                outputStream.write(`ERROR reading file: ${error.message}\n\n`);
                                skippedCount++;
                            }
                        } else {
                            skippedCount++;
                        }
                    }
                } catch (error) {
                    skippedCount++;
                }
            }
        } catch (error) {
            // Skip inaccessible directories
        }
    }
    
    await processDirectory(rootPath);
    return { exportedCount, skippedCount };
}

/**
 * Main export function
 */
async function main() {
    console.log('🚀 Starting MeetingMind codebase export...');
    console.log(`📁 Root directory: ${ROOT_DIR}`);
    console.log(`📄 Output file: ${OUTPUT_FILE}`);
    
    try {
        // Calculate statistics
        console.log('📊 Calculating codebase statistics...');
        const stats = await calculateStats(ROOT_DIR);
        
        // Create output stream
        const outputStream = fs.createWriteStream(OUTPUT_FILE, 'utf8');
        
        // Write header
        outputStream.write('# MeetingMind Complete Codebase Export\n');
        outputStream.write(`Generated: ${new Date().toISOString()}\n`);
        outputStream.write(`Root Path: ${ROOT_DIR}\n\n`);
        
        // Write statistics
        outputStream.write('## Codebase Statistics\n');
        outputStream.write(`- Total Files: ${stats.totalFiles.toLocaleString()}\n`);
        outputStream.write(`- Total Lines: ${stats.totalLines.toLocaleString()}\n`);
        outputStream.write(`- Total Size: ${formatSize(stats.totalSize)}\n`);
        outputStream.write(`- Directories: ${stats.directories.toLocaleString()}\n\n`);
        
        // File types distribution
        outputStream.write('### File Types Distribution\n');
        const sortedFileTypes = Object.entries(stats.fileTypes).sort((a, b) => b[1] - a[1]);
        for (const [ext, count] of sortedFileTypes) {
            const percentage = ((count / stats.totalFiles) * 100).toFixed(1);
            outputStream.write(`- ${ext}: ${count.toLocaleString()} files (${percentage}%)\n`);
        }
        outputStream.write('\n');
        
        // Language distribution
        if (Object.keys(stats.languages).length > 0) {
            outputStream.write('### Language Distribution (by lines)\n');
            const sortedLanguages = Object.entries(stats.languages).sort((a, b) => b[1] - a[1]);
            for (const [lang, lines] of sortedLanguages) {
                const percentage = ((lines / stats.totalLines) * 100).toFixed(1);
                outputStream.write(`- ${lang}: ${lines.toLocaleString()} lines (${percentage}%)\n`);
            }
            outputStream.write('\n');
        }
        
        // Directory structure
        console.log('🌳 Generating directory tree...');
        outputStream.write('## Directory Structure\n```\n');
        outputStream.write(`${path.basename(ROOT_DIR)}/\n`);
        const tree = await getDirectoryTree(ROOT_DIR);
        for (const line of tree) {
            outputStream.write(line + '\n');
        }
        outputStream.write('```\n\n');
        
        // Export configuration
        outputStream.write('## Export Configuration\n');
        outputStream.write('### Included Extensions:\n');
        for (const ext of Array.from(INCLUDE_EXTENSIONS).sort()) {
            outputStream.write(`- ${ext}\n`);
        }
        outputStream.write('\n### Excluded Directories:\n');
        for (const dir of Array.from(EXCLUDE_DIRS).sort()) {
            outputStream.write(`- ${dir}\n`);
        }
        outputStream.write(`\n### Maximum File Size: ${formatSize(MAX_FILE_SIZE)}\n\n`);
        
        // File contents
        outputStream.write('## Complete File Contents\n');
        
        console.log('📄 Exporting file contents...');
        const { exportedCount, skippedCount } = await exportFiles(ROOT_DIR, outputStream);
        
        // Footer
        outputStream.write('\n' + '='.repeat(80) + '\n');
        outputStream.write('# Export Summary\n');
        outputStream.write(`- Files exported: ${exportedCount.toLocaleString()}\n`);
        outputStream.write(`- Files skipped: ${skippedCount.toLocaleString()}\n`);
        outputStream.write(`- Export completed: ${new Date().toISOString()}\n`);
        outputStream.write('='.repeat(80) + '\n');
        
        outputStream.end();
        
        // Final summary
        const outputStats = await stat(OUTPUT_FILE);
        console.log('\n✅ Export completed successfully!');
        console.log('📊 Statistics:');
        console.log(`   - Files exported: ${exportedCount.toLocaleString()}`);
        console.log(`   - Files skipped: ${skippedCount.toLocaleString()}`);
        console.log(`   - Total lines: ${stats.totalLines.toLocaleString()}`);
        console.log(`   - Total size: ${formatSize(stats.totalSize)}`);
        console.log(`📄 Output file: ${OUTPUT_FILE}`);
        console.log(`📏 Output size: ${formatSize(outputStats.size)}`);
        console.log('🎉 Codebase export complete!');
        
    } catch (error) {
        console.error('❌ Export failed:', error);
        process.exit(1);
    }
}

// Run the export
if (require.main === module) {
    main().catch(console.error);
}

module.exports = { main };

File: export_codebase.py
#!/usr/bin/env python3
"""
MeetingMind Codebase Export Script
Exports all code files, structure, and content to a comprehensive text file
"""

import os
import datetime
import fnmatch
from pathlib import Path
import mimetypes

# Configuration
ROOT_DIR = Path(__file__).parent
OUTPUT_FILE = ROOT_DIR / "gemini_codebase_context_full.txt"

# File patterns to include (case-insensitive) - COMPREHENSIVE for complete export
INCLUDE_PATTERNS = [
    "*.ts", "*.tsx", "*.js", "*.jsx", "*.json", "*.yaml", "*.yml",
    "*.py", "*.md", "*.txt", "*.env*", "*.config.*", "*.conf",
    "*.html", "*.css", "*.scss", "*.sass", "*.less",
    "*.sql", "*.sh", "*.bat", "*.ps1", "*.dockerfile", "Dockerfile*",
    "*.gitignore", "*.gitattributes", "*.editorconfig",
    "*.lock", "*.toml", "*.ini", "*.cfg", "*.properties",
    "*.xml", "*.svg", "*.vue", "*.svelte", "*.php", "*.rb",
    "*.go", "*.rs", "*.cpp", "*.c", "*.h", "*.hpp",
    "*.java", "*.kt", "*.swift", "*.dart", "*.r",
    "package.json", "requirements.txt", "Pipfile", "poetry.lock",
    "Cargo.toml", "go.mod", "pom.xml", "build.gradle",
    "tsconfig*.json", "vite.config.*", "webpack.config.*",
    "tailwind.config.*", "postcss.config.*", "babel.config.*",
    "eslint.config.*", ".eslintrc*", ".prettierrc*", "jest.config.*",
    "*.mako", "*.makefile", "Makefile", "*.cmake", "CMakeLists.txt",
    "*.1", "*.example", "*.template", "*.spec", "*.test.*",
    "LICENSE", "COPYING", "README", "CHANGELOG", "VERSION",
    "*.dev", "*.prod", "*.staging", "*.local", "*.backend", "*.frontend"
]

# Directories to exclude
EXCLUDE_DIRS = {
    "node_modules", "__pycache__", ".git", ".vscode", ".idea",
    "dist", "build", "coverage", ".nyc_output", "target",
    ".pytest_cache", ".mypy_cache", ".tox", "venv", "env",
    ".env", "virtualenv", ".virtualenv", "site-packages",
    "vendor", "bower_components", ".next", ".nuxt", ".cache",
    "tmp", "temp", ".tmp", ".temp", "logs", "log",
    ".DS_Store", "Thumbs.db", ".sass-cache", ".parcel-cache",
    "gh_2.74.2_linux_amd64"  # Exclude gh manual pages directory
}

# Files to exclude (specific filenames) - minimal exclusions for complete export
EXCLUDE_FILES = {
    "gemini_codebase_context_full.txt"  # Don't include the output file itself
}

# File extensions to exclude (binary files)
EXCLUDE_EXTENSIONS = {
    ".exe", ".dll", ".so", ".dylib", ".bin", ".obj", ".o",
    ".png", ".jpg", ".jpeg", ".gif", ".bmp", ".ico", ".tiff",
    ".mp3", ".mp4", ".wav", ".avi", ".mov", ".wmv", ".flv",
    ".pdf", ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx",
    ".zip", ".tar", ".gz", ".rar", ".7z", ".bz2", ".xz",
    ".woff", ".woff2", ".ttf", ".eot", ".otf",
    ".db", ".sqlite", ".sqlite3", ".mdb", ".accdb"
}

# Maximum file size to include (in bytes) - NO LIMIT for complete export
MAX_FILE_SIZE = float('inf')  # No limit - include everything

def should_include_file(file_path: Path) -> bool:
    """Check if a file should be included in the export"""
    
    # Check if filename should be excluded
    if file_path.name in EXCLUDE_FILES:
        return False
    
    # Check if file extension should be excluded
    if file_path.suffix.lower() in EXCLUDE_EXTENSIONS:
        return False
    
    # Check file size
    try:
        if file_path.stat().st_size > MAX_FILE_SIZE:
            return False
    except (OSError, IOError):
        return False
    
    # Check if filename matches include patterns
    filename = file_path.name.lower()
    filename_full = file_path.name  # Also check non-lowercase
    for pattern in INCLUDE_PATTERNS:
        if fnmatch.fnmatch(filename, pattern.lower()) or fnmatch.fnmatch(filename_full, pattern):
            return True
    
    # Check if it's a text file by content
    try:
        mime_type, _ = mimetypes.guess_type(str(file_path))
        if mime_type and mime_type.startswith('text/'):
            return True
    except:
        pass
    
    # Be more aggressive - try to read any file that might be text
    try:
        # Skip very large files only if they're clearly binary
        file_size = file_path.stat().st_size
        if file_size > 50 * 1024 * 1024:  # 50MB limit for text detection
            return False
            
        with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
            content_sample = f.read(1000)  # Read more to better detect text
            # Check if content looks like text (has reasonable ratio of printable chars)
            printable_chars = sum(1 for c in content_sample if c.isprintable() or c.isspace())
            if len(content_sample) > 0 and printable_chars / len(content_sample) > 0.7:
                return True
    except (UnicodeDecodeError, IOError, OSError):
        pass
    
    return False

def should_exclude_dir(dir_path: Path) -> bool:
    """Check if a directory should be excluded"""
    dir_name = dir_path.name.lower()
    return dir_name in {d.lower() for d in EXCLUDE_DIRS}

def get_directory_tree(root_path: Path, prefix: str = "") -> str:
    """Generate a visual directory tree structure"""
    tree_lines = []
    
    try:
        items = sorted(root_path.iterdir(), key=lambda x: (not x.is_dir(), x.name.lower()))
        
        for i, item in enumerate(items):
            if should_exclude_dir(item) and item.is_dir():
                continue
                
            is_last = i == len(items) - 1
            current_prefix = "└── " if is_last else "├── "
            tree_lines.append(f"{prefix}{current_prefix}{item.name}")
            
            if item.is_dir() and not should_exclude_dir(item):
                extension = "    " if is_last else "│   "
                subtree = get_directory_tree(item, prefix + extension)
                if subtree:
                    tree_lines.append(subtree)
    
    except PermissionError:
        pass
    
    return "\n".join(tree_lines)

def get_file_stats(root_path: Path) -> dict:
    """Calculate statistics about the codebase"""
    stats = {
        'total_files': 0,
        'total_lines': 0,
        'total_size': 0,
        'file_types': {},
        'directories': 0,
        'languages': {}
    }
    
    for root, dirs, files in os.walk(root_path):
        root_path_obj = Path(root)
        
        # Exclude certain directories
        dirs[:] = [d for d in dirs if not should_exclude_dir(root_path_obj / d)]
        
        stats['directories'] += 1
        
        for file in files:
            file_path = root_path_obj / file
            
            if not should_include_file(file_path):
                continue
            
            try:
                file_size = file_path.stat().st_size
                stats['total_files'] += 1
                stats['total_size'] += file_size
                
                # Count file types
                ext = file_path.suffix.lower()
                if ext:
                    stats['file_types'][ext] = stats['file_types'].get(ext, 0) + 1
                else:
                    stats['file_types']['no_extension'] = stats['file_types'].get('no_extension', 0) + 1
                
                # Count lines
                try:
                    with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
                        lines = sum(1 for _ in f)
                        stats['total_lines'] += lines
                        
                        # Language detection
                        if ext in ['.ts', '.tsx']:
                            stats['languages']['TypeScript'] = stats['languages'].get('TypeScript', 0) + lines
                        elif ext in ['.js', '.jsx']:
                            stats['languages']['JavaScript'] = stats['languages'].get('JavaScript', 0) + lines
                        elif ext == '.py':
                            stats['languages']['Python'] = stats['languages'].get('Python', 0) + lines
                        elif ext in ['.css', '.scss', '.sass', '.less']:
                            stats['languages']['CSS'] = stats['languages'].get('CSS', 0) + lines
                        elif ext in ['.html', '.htm']:
                            stats['languages']['HTML'] = stats['languages'].get('HTML', 0) + lines
                        elif ext in ['.md']:
                            stats['languages']['Markdown'] = stats['languages'].get('Markdown', 0) + lines
                        elif ext in ['.json']:
                            stats['languages']['JSON'] = stats['languages'].get('JSON', 0) + lines
                        elif ext in ['.yaml', '.yml']:
                            stats['languages']['YAML'] = stats['languages'].get('YAML', 0) + lines
                        else:
                            stats['languages']['Other'] = stats['languages'].get('Other', 0) + lines
                            
                except (UnicodeDecodeError, IOError):
                    pass
                    
            except (OSError, IOError):
                pass
    
    return stats

def format_size(size_bytes: int) -> str:
    """Format file size in human readable format"""
    for unit in ['B', 'KB', 'MB', 'GB']:
        if size_bytes < 1024:
            return f"{size_bytes:.1f} {unit}"
        size_bytes /= 1024
    return f"{size_bytes:.1f} TB"

def export_file_content(file_path: Path, output_file) -> bool:
    """Export the content of a single file"""
    try:
        relative_path = file_path.relative_to(ROOT_DIR)
        file_size = file_path.stat().st_size
        
        output_file.write(f"\n{'='*80}\n")
        output_file.write(f"FILE: {relative_path}\n")
        output_file.write(f"SIZE: {format_size(file_size)}\n")
        output_file.write(f"MODIFIED: {datetime.datetime.fromtimestamp(file_path.stat().st_mtime)}\n")
        output_file.write(f"{'='*80}\n\n")
        
        with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
            content = f.read()
            
            # Add line numbers for ALL text files (not just specific types)
            lines = content.split('\n')
            numbered_content = []
            for i, line in enumerate(lines, 1):
                numbered_content.append(f"{i:5d}→ {line}")
            content = '\n'.join(numbered_content)
            
            output_file.write(content)
            output_file.write("\n\n")
        
        return True
        
    except (UnicodeDecodeError, IOError, OSError) as e:
        output_file.write(f"ERROR reading file {file_path}: {e}\n\n")
        return False

def main():
    """Main export function"""
    print(f"🚀 Starting MeetingMind codebase export...")
    print(f"📁 Root directory: {ROOT_DIR}")
    print(f"📄 Output file: {OUTPUT_FILE}")
    
    # Calculate statistics
    print("📊 Calculating codebase statistics...")
    stats = get_file_stats(ROOT_DIR)
    
    # Start writing the export file
    with open(OUTPUT_FILE, 'w', encoding='utf-8') as output:
        # Header
        output.write("# MeetingMind Codebase Context for Gemini\n")
        output.write(f"Generated: {datetime.datetime.now().isoformat()}\n")
        output.write(f"Root Path: {ROOT_DIR}\n")
        output.write("\n")
        
        # Statistics
        output.write("## Codebase Statistics\n")
        output.write(f"- Total Files: {stats['total_files']:,}\n")
        output.write(f"- Total Lines: {stats['total_lines']:,}\n")
        output.write(f"- Total Size: {format_size(stats['total_size'])}\n")
        output.write(f"- Directories: {stats['directories']:,}\n")
        output.write("\n")
        
        # File types breakdown
        output.write("### File Types Distribution\n")
        for ext, count in sorted(stats['file_types'].items(), key=lambda x: x[1], reverse=True):
            percentage = (count / stats['total_files']) * 100
            output.write(f"- {ext}: {count:,} files ({percentage:.1f}%)\n")
        output.write("\n")
        
        # Language breakdown
        if stats['languages']:
            output.write("### Language Distribution (by lines)\n")
            for lang, lines in sorted(stats['languages'].items(), key=lambda x: x[1], reverse=True):
                percentage = (lines / stats['total_lines']) * 100
                output.write(f"- {lang}: {lines:,} lines ({percentage:.1f}%)\n")
            output.write("\n")
        
        # Directory structure
        output.write("## Directory Structure\n")
        output.write("```\n")
        output.write(f"{ROOT_DIR.name}/\n")
        tree = get_directory_tree(ROOT_DIR)
        output.write(tree)
        output.write("\n```\n\n")
        
        # Export configuration
        output.write("## Export Configuration\n")
        output.write("### Included File Patterns:\n")
        for pattern in INCLUDE_PATTERNS:
            output.write(f"- {pattern}\n")
        output.write("\n")
        
        output.write("### Excluded Directories:\n")
        for dir_name in sorted(EXCLUDE_DIRS):
            output.write(f"- {dir_name}\n")
        output.write("\n")
        
        output.write("### Excluded Extensions:\n")
        for ext in sorted(EXCLUDE_EXTENSIONS):
            output.write(f"- {ext}\n")
        output.write("\n")
        
        output.write(f"### Maximum File Size: {format_size(MAX_FILE_SIZE)}\n\n")
        
        # File contents
        output.write("## Complete File Contents\n\n")
        
        # Walk through all files and export content
        exported_count = 0
        skipped_count = 0
        
        for root, dirs, files in os.walk(ROOT_DIR):
            root_path = Path(root)
            
            # Exclude certain directories
            dirs[:] = [d for d in dirs if not should_exclude_dir(root_path / d)]
            
            for file in sorted(files):
                file_path = root_path / file
                
                if not should_include_file(file_path):
                    skipped_count += 1
                    continue
                
                print(f"📝 Exporting: {file_path.relative_to(ROOT_DIR)}")
                
                if export_file_content(file_path, output):
                    exported_count += 1
                else:
                    skipped_count += 1
        
        # Footer
        output.write("\n" + "="*80 + "\n")
        output.write("# Export Summary\n")
        output.write(f"- Files exported: {exported_count:,}\n")
        output.write(f"- Files skipped: {skipped_count:,}\n")
        output.write(f"- Export completed: {datetime.datetime.now().isoformat()}\n")
        output.write("="*80 + "\n")
    
    # Final summary
    print(f"\n✅ Export completed successfully!")
    print(f"📊 Statistics:")
    print(f"   - Files exported: {exported_count:,}")
    print(f"   - Files skipped: {skipped_count:,}")
    print(f"   - Total lines: {stats['total_lines']:,}")
    print(f"   - Total size: {format_size(stats['total_size'])}")
    print(f"📄 Output file: {OUTPUT_FILE}")
    print(f"📏 Output size: {format_size(OUTPUT_FILE.stat().st_size)}")

if __name__ == "__main__":
    main()